add filter for intermediate resonance particles

jieunyoo opened the pull request:
Add this GenFilter. There is a similar LHEPtFilter, but it only filters outgoing particles. This filter does the filtering on intermediate particles. For example, for a boosted Higgs, we would like to filter on the pT of the Higgs (not the final state partons). The current LHEPtFilter does not do that. This doesn't affect any other PRs or external code. 

(An attempt to extend the current LHEPtFilter was made instead of adding this filter, but difficulties were encountered with the plugin and I hope to get this addition added soon so we can ask for a small production of the sample.)

I would like to backport this also to 10_6_X to that it can be used for a Run 2 UL thesis.

jieunyoo committed:
add filter for intermediate resonance particles

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33305

- This PR adds an extra 12KB to repository

cmsbuild commented:
A new Pull Request was created by  @jieunyoo (jieun) for master.

It involves the following packages:

- GeneratorInterface/GenFilters (**generators**)


@SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please review it and eventually sign? Thanks.
@alberto-sanchez, @mkirsano this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

jieunyoo commented:
@cmsbuild please test

jieunyoo commented:
@menglu21 Thanks!

jieunyoo commented:
This filter works simply like this:

LHEHiggsPtFilter = cms.EDFilter("LHEIntermediateParticlePtFilter",
  selectedPdgIds = cms.vint32(25),
  ptMin=cms.double(135.),
  ptMax=cms.double(1e10),
  src=cms.InputTag("externalLHEProducer")
  
  A small private sample was made. This shows the filtered pT of the Higgs.
![higgsPT-filtered](https://user-images.githubusercontent.com/23608596/206583447-d0c4fee5-e605-4b83-90fe-d8b196a1afc4.png)

menglu21 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-85908a/29555/summary.html
**COMMIT**: d4d58b687b2515d56abfa5914f29b48aac6612c4
**CMSSW**: CMSSW_13_0_X_2022-12-11-0000/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40254/29555/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 33 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421510
 - DQMHistoTests: Total failures: 1192
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3420296
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-11-0000+85908a/54607/triggerResults): no differences found

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta reviewed (commented):
Since this is a new filter, please also add a fillDescriptions method to define and validate the parameters

jieunyoo committed:
Update GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

jieunyoo committed:
Update GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

jieunyoo committed:
Update GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

Update GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

Update GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

addresses comments

Code suggestion introduced an errant parenthesis which led this script to fail code checks; this was fixed. Comments on the implementation are implemented to address request.

previous patch wasn't applied. try again

curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33349/code-format.patch | patch -p1

remove iostream

addresses comments

Code suggestion introduced an errant parenthesis which led this script to fail code checks; this was fixed. Comments on the implementation are implemented to address request.

previous patch wasn't applied. try again

curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33349/code-format.patch | patch -p1

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33343

ERROR: Build errors found during clang-tidy run.
```
GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc:95:1: error: extraneous closing brace ('}') [clang-diagnostic-error]
}
^
Suppressed 1094 warnings (1093 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33347

- This PR adds an extra 12KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33347/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33347/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33349

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33349/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33349/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

jieunyoo commented on GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc:
```diff
@@ -0,0 +1,105 @@
+
+// system include files
+#include <memory>
+#include <iostream>
```
removed iostream; other filters have memory so kept it

jieunyoo replied:
From checking other filters, it seems they had an "implementation section" in the comments. I added that section. Hope that addresses the above "fillDescriptions method." (I didn't see any specific fillDescriptions example.) Thanks for the quick review!

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33350

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40254 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

perrotta commented:
@jieunyoo 
- you did not remove `iostream` in fact...
- fillDescriptions is a method used to validate and initialize the configuration parameters, see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp . If you don't know how to implement it, don't mind: it is a rather simple filter, we can even skip validation...
- At the end of your fixes, please squash all commits into one so that you can clean up the history of the file

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40254/33358

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #40254 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

jieunyoo commented:
Hi @perrotta . I squashed the commits, and removed "iostream." For this commit then (since it's my first ever to CMSSW), I guess I will skip the fillDescriptions method that you mentioned above (you mentioned it would be ok to skip). Thanks for the reference though and I'll keep it in mind for later. Does everything look good for you? Do I need to do anything else for a backport to 10_6_X for UL studies? I am not sure of the timeline. Hope to get this to my MC contact for a UL MC production. Thanks!

perrotta commented:
Please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-85908a/29578/summary.html
**COMMIT**: a011ddf507aeb4de2631319ce603d1d8c01c7c9e
**CMSSW**: CMSSW_13_0_X_2022-12-12-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40254/29578/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421510
 - DQMHistoTests: Total failures: 152
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3421336
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-12-1100+85908a/54630/triggerResults): no differences found

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-85908a/29590/summary.html
**COMMIT**: a011ddf507aeb4de2631319ce603d1d8c01c7c9e
**CMSSW**: CMSSW_13_0_X_2022-12-12-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40254/29590/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421510
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3421482
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-12-2300+85908a/54642/triggerResults): no differences found

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
